pkgs/by-name/av/avalonia/package.nix

```diff
@@ -0,0 +1,175 @@
+            --replace-fail '"libXrandr.so.2"' '"${lib.getLib libXrandr}/lib/libXrandr.so.2"' \
+            --replace-fail '"libXext.so.6"' '"${lib.getLib libXext}/lib/libXext.so.6"' \
+            --replace-fail '"libXi.so.6"' '"${lib.getLib libXi}/lib/libXi.so.6"' \
+            --replace-fail '"libXcursor.so.1"' '"${lib.getLib libXcursor}/lib/libXcursor.so.1"'
+
+          # from RestoreAdditionalProjectSources, which isn't supported by nuget-to-nix
+          dotnet nuget add source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8-transport/nuget/v3/index.json
+
+          # Tricky way to run npmConfigHook multiple times (borrowed from pagefind)
+          (
+            local postPatchHooks=() # written to by npmConfigHook
+            source ${npmHooks.npmConfigHook}/nix-support/setup-hook
+        ''
+        +
+          lib.concatMapStrings
+            (
+              { path, hash }:
+              let
+                deps = fetchNpmDeps {
+                  src = "${src}/${path}";
+                  inherit hash;
+                };
+              in
+              ''
+                npmRoot=${path} npmDeps="${deps}" npmConfigHook
+                rm -rf "$TMPDIR/cache"
+              ''
+            )
+            [
+              {
+                path = "src/Avalonia.DesignerSupport/Remote/HtmlTransport/webapp";
+                hash = "sha256-gncHW5SMtAUMtvHGZ2nUc0KEjxX24DZkAnmeHgo1Roc=";
+              }
+              {
+                path = "tests/Avalonia.DesignerSupport.Tests/Remote/HtmlTransport/webapp";
+                hash = "sha256-MiznlOJ+hIO1cUswy9oGNHP6MWfx+FDLKVT8qcmg8vo=";
+              }
+              {
+                path = "src/Browser/Avalonia.Browser/webapp";
+                hash = "sha256-LTQzT4ycLyGQs9T0sa2k/0wfG1GWCdeH9Wx2KeecOyU=";
+              }
+            ]
+        + ''
+          )
+        '';
+
+      makeCacheWritable = true;
+
+      # CSC : error CS1566: Error reading resource 'pdbstr.exe' -- 'Could not find a part of the path '/build/.nuget-temp/packages/sourcelink/1.1.0/tools/pdbstr.exe'.' [/build/source/nukebuild/_build.csproj]
+      linkNugetPackages = true;
+
+      # [WRN] Could not inject value for Build.ApiCompatTool
+      # System.Exception: Missing package reference/download.
+      # Run one of the following commands:
+      #  ---> System.ArgumentException: Could not find package 'Microsoft.DotNet.ApiCompat.Tool' using:
+      #  - Project assets file '/build/source/nukebuild/obj/project.assets.json'
+      #  - NuGet packages config '/build/source/nukebuild/_build.csproj'
+      makeEmptyNupkgInPackages = true;
+
+      buildInputs = [
+        (
+          with dotnetCorePackages;
+          combinePackages [
+            sdk_6_0
+            sdk_7_0_1xx
+          ]
+        )
+        libX11
+        libICE
+        libSM
+        libXcursor
+        libXrandr
+        libXext
+        libXi
```

Comment:
Should any of these be _propagated_ inputs?

I'm not very confident in knowing which to use when, but if `avalonia` is itself a dependency of other packages, should these be propagated as transitive dependencies to said packages? Assuming that's even what `propagatedBuildInputs` does

Reply:
They might not actually need to be `buildInputs` at all.  Avalonia just `dlopen`s these from string literals, so it's patching in the full paths.

```
          substituteInPlace src/Avalonia.X11/ICELib.cs \
            --replace-fail '"libICE.so.6"' '"${lib.getLib libICE}/lib/libICE.so.6"'
          substituteInPlace src/Avalonia.X11/SMLib.cs \
            --replace-fail '"libSM.so.6"' '"${lib.getLib libSM}/lib/libSM.so.6"'
          substituteInPlace src/Avalonia.X11/XLib.cs \
            --replace-fail '"libX11.so.6"' '"${lib.getLib libX11}/lib/libX11.so.6"' \
            --replace-fail '"libXrandr.so.2"' '"${lib.getLib libXrandr}/lib/libXrandr.so.2"' \
            --replace-fail '"libXext.so.6"' '"${lib.getLib libXext}/lib/libXext.so.6"' \
            --replace-fail '"libXi.so.6"' '"${lib.getLib libXi}/lib/libXi.so.6"' \
            --replace-fail '"libXcursor.so.1"' '"${lib.getLib libXcursor}/lib/libXcursor.so.1"'
```

Maybe there's a cleverer way to do something like RPATH with `DllImport`?

I do also need to check that these dependencies get picked up by nix, since they're probably in a UTF16 string somewhere.  I'm sure if nix can handle that.

Reply:
I've now removed them from `buildInputs`.